feat: Serde Serialize and Deserialize traits for the RbTree in the ic-certified-map crate.

levifeldman opened the pull request:
This change adds the CandidType, serde Serialize and Deserialize traits to the RbTree in the ic-certified-map crate.
Useful for serializing a canister's certified data during upgrades.

This PR changes the 'static generic lifetime bounds on the RbTree impl blocks and replaces it with a custom lifetime 't.

# How Has This Been Tested?

library/ic-certified-map/src/rbtree/test.rs

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

levifeldman commented:
Hey, I replaced the 'static generic lifetime bounds on the RbTree impl blocks for a custom lifetime 't.

levifeldman commented:
and added a test

levifeldman commented:
Looks like the new canister import functionality is messing with the PR check: https://github.com/dfinity/cdk-rs/actions/runs/6156757139/job/16706101327?pr=399

levifeldman commented:
Looks like the new canister import functionality is messing with the PR check: https://github.com/dfinity/cdk-rs/actions/runs/6156757139/job/16706101327?pr=399

adamspofford-dfinity commented on library/ic-certified-map/src/rbtree.rs:
```diff
@@ -888,6 +888,95 @@ impl<K: 'static + AsRef<[u8]>, V: AsHashTree + 'static> RbTree<K, V> {
+    where
+        S: Serializer,
+    {
+        let collect_refs = self.iter().collect::<Vec<(&K, &V)>>();
```
Why is the iterator being buffered into a vector instead of serialized directly?

levifeldman replied:
Some binary formats like bincode need to know the length of the map upfront and that gets passed to the `serialize_map` function on the next line. The comments here say more: https://docs.rs/serde/latest/serde/ser/trait.Serializer.html#tymethod.serialize_map.

levifeldman replied:
Now in commit https://github.com/dfinity/cdk-rs/pull/399/commits/a1779b4adaf31db6c85aed6090af72b58797aabe it uses self.iter().count() instead of allocating a vector.

levifeldman commented:
Hey @adamspofford-dfinity, LGTM.

levifeldman commented:
Thanks Adam @adamspofford-dfinity for the review and merge. 

My apologies for putting off addressing your feedback for a while there. Won't happen again.
